signing/vrf.go

```diff
@@ -10,26 +10,32 @@ import (
 
 	"github.com/spacemeshos/go-spacemesh/common/types"
 	"github.com/spacemeshos/go-spacemesh/datastore"
-	"github.com/spacemeshos/go-spacemesh/sql/atxs"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/sql/vrfnonce"
 )
 
 // nonceFetcher is an abstraction for VRFSigner and VRFVerifier to fetch a nonce for a given node.
 type nonceFetcher interface {
-	NonceForNode(types.NodeID) (types.VRFPostIndex, error)
+	NonceForNode(types.NodeID, types.EpochID) (types.VRFPostIndex, error)
 }
 
 // VRFSigner is a signer for VRF purposes.
 type VRFSigner struct {
-	privateKey []byte
-	fetcher    nonceFetcher
+	fetcher nonceFetcher
+	log     log.Log
 
-	nodeID types.NodeID
+	privateKey []byte
+	nodeID     types.NodeID
 }
 
 // Sign signs a message for VRF purposes.
-func (s VRFSigner) Sign(msg []byte) ([]byte, error) {
-	nonce, err := s.fetcher.NonceForNode(s.nodeID)
+func (s VRFSigner) Sign(msg []byte, epoch types.EpochID) ([]byte, error) {
+	nonce, err := s.fetcher.NonceForNode(s.nodeID, epoch)
```

Comment:
this API is confusing.

it is more intuitive to pass the nonce as a parameter to Sign, instead of passing all the parameters needed to find the nonce.

in terms of code dependency, it is strange that we need a database to create a vrf signer. it should be just a nonce, regardless of where the nonce is stored.

besides, and it's prolly sounding old now, that i still think the code will be easier to read if we just add nonce to be part of the vrf message.

Reply:
I chose this design for the following reason: Everywhere `VRFSigner::Sign` is used (and similarly for `VRFVerifier::Verify`), the epoch is known and can be passed into the function.

If instead `Sign` (or `Verify`) would take the Nonce as parameter all places where they are invoked need to be updated to fetch the nonce first using the `NodeID` (that in the places where `Sign` is used, isn't readily available) and `epoch` just to then pass it into this function.

> besides, and it's prolly sounding old now, that i still think the code will be easier to read if we just add nonce to be part of the vrf message.

We had the same discussion in the last PR with VRF changes 🙂, if you want instead of concatenating `msg` and `nonce` I can use scale to encode them into the following struct:

```golang
type VRFMessage {
    Nonce   types.VRFPostIndex
    Message []byte
}
```

and use that to encode / decode the byte array that is signed or verified against.

Reply:
this is a huge mistake in my opinion, it just accidentally adds fetcher and database to the hare codepath, it changes latency expectations by quite a lot

Reply:
for the record, this is not what i suggested. take beacon vrf message for example, it would result in the following
```
type beaconVRF struct {
  beacon_eligibility
  nonce
  epoch  
}
```
that way vrf verifier does not need to take any parameter at all. it's up to the app logic to construct the expected vrf message and verify the proof (vrf signature) against it.